File: python/examples/02_gift_leaderboard.py
"""Live gift leaderboard. Refreshes every 5s.

Run:
    USERNAME=streamer python 02_gift_leaderboard.py
"""
import asyncio
import os
from collections import defaultdict
from tiktok_live_events import TikTokLive

live = TikTokLive(os.environ.get("USERNAME", "tiktokuser"))
board: dict[str, int] = defaultdict(int)


@live.on("gift")
def on_gift(e):
    if not e.get("repeatEnd"):
        return
    board[e["user"]["uniqueId"]] += e["diamondCount"] * e["repeatCount"]


async def print_top():
    while True:
        await asyncio.sleep(5)
        top = sorted(board.items(), key=lambda x: -x[1])[:10]
        print("\n-- TOP GIFTERS --")
        for i, (user, diamonds) in enumerate(top, 1):
            print(f"{i:>2}. {user:<20} {diamonds} diamonds")


async def main():
    asyncio.create_task(print_top())
    await live.run()


asyncio.run(main())

File: node/examples/02-gift-leaderboard.js
// Live gift leaderboard. Refreshes every 5s.
// Run: USERNAME=streamername node 02-gift-leaderboard.js
import { TikTokLive } from 'tiktok-live-events';

const live = new TikTokLive(process.env.USERNAME || 'tiktokuser');
const board = new Map();

live.on('gift', e => {
    if (!e.repeatEnd) return;
    const total = (board.get(e.user.uniqueId) || 0) + e.diamondCount * e.repeatCount;
    board.set(e.user.uniqueId, total);
});

setInterval(() => {
    console.clear();
    console.log('-- TOP GIFTERS --');
    [...board.entries()]
        .sort((a, b) => b[1] - a[1])
        .slice(0, 10)
        .forEach(([u, d], i) => console.log(`${i + 1}. ${u.padEnd(20)} ${d} diamonds`));
}, 5_000);

await live.connect();
